graphs/wolf_goat_cabbage.py

```diff
@@ -0,0 +1,159 @@
+
+    def __int__(self) -> int:
+        """
+        Maps the state to a unique key based on contents. The things on the left side
+        are added to the key, and the things on the right side are subtracted from the
+        key.
+        
+        Returns:
+            int: The value of the key
+        """
+        key = 0
+        for thing in self.leftSide:
+            key += State.value[thing]
+        
+        for thing in self.rightSide:
+            key -= State.value[thing]
+        return key
+
+    def __str__(self) -> str:
+        return "State [left side: {}, right side: {}".format(self.leftSide, self.rightSide)
+
+    @staticmethod
+    def __moveFarmerAndThing(state, thing) -> List['State']:
+        """
+        Moves farmer and a thing from one side to another.
+        
+        Returns:
+            List['State']: A list containing the possible states that result in 
+            such movement.
+        """
+
+        # if the farmer is on the left side, run if body
+        if 'f' in state.leftSide:
+            # generate a list containing all things except the farmer and `thing` for 
+            # the left side
+            leftSide = [i for i in state.leftSide if i != 'f' and i != thing]
+            # make a list containing the right side things, the farmer and `thing`
+            rightSide = state.rightSide + ['f', thing]
+            # create a state object and return it
+            return [State(left=leftSide, right=rightSide)]
+
+        # this code runs when the farmer is on the right side. When the farmer
+        # is on the right side, he can either come back with a thing or come back
+        # alone. Hence there are two possible states.
+
+        # variable that stores the possible states
+        states = []
+        
+        # when the farmer comes back with a thing:
+
+        # generate list containing all things except the farmer and `thing` for the
+        # right side
+        rightSide = [i for i in state.rightSide if i != 'f' and i != thing]
+
+        # create a state with the generated list for the right side, and left side
+        # things + farmer and `thing` for the left side
+        states.append(State(left=state.leftSide + ['f', thing], right=rightSide))
+
+        # when farmer comes back alone:
+
+        # generate a list containing all things except farmer for the right side
+        rightSide = [i for i in state.rightSide if i != 'f']
+
+        # create a state with the generated list for the right side and left side
+        # things + farmer for the left side
+        states.append(State(left=state.leftSide + ['f'], right=rightSide))
+
+        # return the states
+        return states
+    
+        
+
+    @staticmethod
+    def initial() -> 'State':
+        """
+        Returns the initial state.
+        
+        Returns:
+            State: The initial state.
+        """
+
+        return State(left=['f', 'g', 'c', 'w'], right=[])
+
+
+if __name__ == "__main__":
```

Comment:
@mitrish refactor the code in your PR, so that it can be run by anyone trying to copy the code